### PR TITLE
reactive streams

### PR DESCRIPTION
**Pull request by tanner0101**
Adds back pressure making vapor's sync streams [reactive](http://www.reactive-streams.org).

### CODE REVIEW

File: Sources/Async/Streams/DeltaStream.swift [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+///         req.requestOutput()
+///     }
+///
+///     numberEmitter.emit(1)
+///     numberEmitter.emit(2)
+///     numberEmitter.emit(3)
+///
+///     print(output) /// [1, 1, 2, 2, 3, 3]
+///
+public final class DeltaStream<Splitting>: Stream {
```
Comment (Joannis) [inline Sources/Async/Streams/DeltaStream.swift]: Is this still correct? The docs probably aren't

File: Sources/Async/Streams/OutputRequest.swift [hunk 2]
```diff
@@ -0,0 +1,34 @@
+/// A `OutputRequest` represents a one-to-one lifecycle of a `InputStream` subscribing to a `OutputStream`.
+///
+/// It can only be used once by a single `InputStream`.
+///
+/// It is used to both signal desire for data and cancel demand (and allow resource cleanup).
+public protocol OutputRequest {
```
Comment (Joannis) [inline Sources/Async/Streams/OutputRequest.swift]: Should this be `: class` ?

File: Sources/Async/Streams/Stream.swift [hunk 3]
```diff
@@ -0,0 +1,3 @@
+/// A `Stream` represents a processing stage—which is both a `InputStream`
+/// and a `OutputStream` and obeys the contracts of both.
+public protocol Stream: InputStream, OutputStream { }
```
Comment (Joannis) [inline Sources/Async/Streams/Stream.swift]: Isn't this better as `public typealias Stream = InputStream & OutputStream`?

**Review by Joannis — APPROVED** [review]
Absolutely *brilliant* improvements. I'd prefer to see this integrated in a "real" situation ( vapor/vapor ) before we can be sure this is the right way forward. Especially considering the impact.

Loving it!

**Comment by tanner0101** [comment]
moving to #15

**Pull request closed without merging**